lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,233 @@
+                                         clonedOp->getResults());
+
+    // Declare the function prototype in the source module
+    localFunc =
+        func::FuncOp::create(opToHoist->getLoc(), functionName, funcType);
+    localFunc.setPrivate(); // Mark as external (no body)
+    sourceModule.push_back(localFunc);
+
+    hoistedFunc->setAttr("arg_ranks", builder.getI64ArrayAttr(ranks));
+  }
+
+  // Replace the original operation with a call to the hoisted function
+  mlir::OpBuilder opBuilder(opToHoist);
+  auto callOp = opBuilder.create<mlir::func::CallOp>(
+      opToHoist->getLoc(), localFunc, opToHoist->getOperands());
+
+  // Replace all results of the original operation with the call results
+  for (auto result : llvm::zip(opToHoist->getResults(), callOp.getResults())) {
+    std::get<0>(result).replaceAllUsesWith(std::get<1>(result));
+  }
+
+  // Erase the original operation
+  opToHoist->erase();
+}
+
+// This pass will collect the operations to be hoisted
+class TTIRHoistAnalyze {
+public:
+  // Custom data structure to store the operations to hoist
+  using HoistOpSet = std::vector<std::set<mlir::Operation *>>;
+
+  // Run the pass and collect operations to hoist
+  TTIRHoistAnalyze(mlir::Operation *op) {
+    auto moduleOp = llvm::dyn_cast<mlir::ModuleOp>(op);
+    assert(moduleOp != nullptr &&
+           "somehow got non-ModuleOp in TTIRHoistAnalyze!");
+
+    // Traverse the operations and look for specific op types
+    // Traverse the module's operations recursively
+    moduleOp.walk([&](mlir::Operation *nestedOp) {
+      if (llvm::isa<ttir::AddOp>(nestedOp) ||
+          llvm::isa<ttir::SubtractOp>(nestedOp) ||
+          llvm::isa<ttir::MultiplyOp>(nestedOp)) {
+        // Store the operations to hoist in the set
+        std::set<Operation *> opSet;
+        opSet.insert(nestedOp);
+        hoistedOps.push_back(opSet);
+      }
+    });
+  }
+
+  HoistOpSet getResults() { return hoistedOps; }
+
+private:
+  HoistOpSet hoistedOps;
+};
+
+class TTIRHoistTransform
+    : public impl::TTIRHoistTransformBase<TTIRHoistTransform> {
+public:
+  using impl::TTIRHoistTransformBase<
+      TTIRHoistTransform>::TTIRHoistTransformBase;
+
+  void runOnOperation() final {
+    const auto &hoistOpSets = getAnalysis<TTIRHoistAnalyze>().getResults();
+
+    mlir::ModuleOp moduleOp = llvm::dyn_cast<mlir::ModuleOp>(getOperation());
+    assert(moduleOp != nullptr && "TTIRHoistTransform should run on ModuleOps, "
+                                  "somehow got something else!");
+    IRRewriter rewriter(&getContext());
+
+    auto loc = moduleOp->getLoc();
+
+    // Check if a "cpu_module" already exists
+    mlir::ModuleOp cpuModule;
+    for (auto &op : moduleOp.getBody()->getOperations()) {
+      if (auto module = llvm::dyn_cast<mlir::ModuleOp>(op)) {
+        if (module->hasAttr("ttir.cpu_module")) {
+          cpuModule = module;
+          break;
+        }
+      }
+    }
+
+    // If no CPU module exists, create one
```

Comment:
What is the intention of splitting the IR into two different modules? Do we need this split? Currently, we have a very simple IR structure with one module and many functions, hence all our passes are working on a single module.

Reply:
Well, the hoisted funcs will be lowered via a separate pipeline, so I don't think we want any of the existing passes running on them.  @nsmithtt originally proposed splitting CPU funcs into new module I think

Reply:
Yeah the intention for splitting into a new module is that we can have a 1-1 mapping of module to device target type.  E.g. the main module is tagged with a system desc, i.e. all the functions within that module are intended to run on device/TTNN framework, a CPU module is tagged with a CPU target triple so organizationally it's a bit easier to read / discover that all functions inside that module are intended for CPU.

On top of that, there are two other reasons:
- The logic for lowering to llvm / compiling to dylib becomes much simpler because we just hand over the CPU module op as the root node for conversion, whereas if we had CPU and TTNN entry point functions just freely together in the same module, we'd have to run a cleanup pass and remove the TTNN functions so that LLVM lowering passes don't run on them.
- As far as we can tell, iree does it this way too.